hawkbit-repository/hawkbit-repository-jpa/src/test/java/org/eclipse/hawkbit/repository/jpa/ControllerManagementTest.java

```diff
@@ -989,4 +1001,154 @@ public void createActionStatusWithTooManyMessages() {
+            "DOWNLOAD_ONLY action.")
+    public void quotaExceptionWhencontrollerReportsTooManyDownloadedMessagesForDownloadOnlyAction() {
+        final int maxMessages = quotaManagement.getMaxMessagesPerActionStatus();
+        testdataFactory.createTarget();
+        final Long actionId = createAndAssignDsAsDownloadOnly(TestdataFactory.DEFAULT_CONTROLLER_ID);
+        assertThat(actionId).isNotNull();
+
+        IntStream.range(0, maxMessages).forEach(i -> controllerManagement
+                .addUpdateActionStatus(entityFactory.actionStatus().create(actionId).status(Status.DOWNLOADED)));
+    }
+
+    @Test
+    @Description("Verifies that no quota exceptions are thrown when a controller reports too many FINISHED events for "
+            + "a DOWNLOAD_ONLY action.")
+    public void noQuotaExceptionWhenControllerReportsTooManyFinishedMessagesForDownloadOnlyAction() {
+        final int maxMessages = quotaManagement.getMaxMessagesPerActionStatus();
+        testdataFactory.createTarget();
+        final Long actionId = createAndAssignDsAsDownloadOnly(TestdataFactory.DEFAULT_CONTROLLER_ID);
+        assertThat(actionId).isNotNull();
+
+        try {
+            IntStream.range(0, maxMessages).forEach(i -> controllerManagement
+                    .addUpdateActionStatus(entityFactory.actionStatus().create(actionId).status(Status.FINISHED)));
+        } catch (QuotaExceededException e) {
+            fail("No QuotaExceededException should be thrown in this context!");
+        }
+
+        assertActionStatus(actionId, TestdataFactory.DEFAULT_CONTROLLER_ID, TargetUpdateStatus.IN_SYNC,
+                Status.FINISHED, Status.FINISHED, false);
+
+        assertThat(actionStatusRepository.count()).isEqualTo(2);
+    }
+
+    @Test
+    @Description("Verifies that a controller can report a FINISHED event for a DOWNLOAD_ONLY action.")
+    public void controllerReportsFinishedForOldDownloadOnlyActionAfterSuccessfulForcedAssignment() {
+        // assign DS as download only assignment
+        testdataFactory.createTarget();
+        final Long downloadOnlyActionId = createAndAssignDsAsDownloadOnly(TestdataFactory.DEFAULT_CONTROLLER_ID);
+        assertThat(downloadOnlyActionId).isNotNull();
+        controllerManagement.addUpdateActionStatus(entityFactory.actionStatus().create(downloadOnlyActionId)
+                        .status(Status.DOWNLOADED));
+        assertActionStatus(downloadOnlyActionId, TestdataFactory.DEFAULT_CONTROLLER_ID, TargetUpdateStatus.IN_SYNC,
+                Status.DOWNLOADED, Status.DOWNLOADED, false);
+        assertThat(actionRepository.activeActionExistsForControllerId(TestdataFactory.DEFAULT_CONTROLLER_ID))
+                .isEqualTo(false);
+        Optional<Action> action = actionRepository.getById(downloadOnlyActionId);
+        Optional<Target> target = targetRepository.findByControllerId(TestdataFactory.DEFAULT_CONTROLLER_ID);
```

Comment:
The code might be more readable if dealing with the Optionals would be transferred into separate methods that eventually safely fail the test (Action getActionFromRepo(id))